pkg/repository/v1manifest/key_store.go

```diff
@@ -0,0 +1,137 @@
+// limitations under the License.
+
+package v1manifest
+
+import (
+	"fmt"
+
+	"github.com/pingcap-incubator/tiup/pkg/repository/crypto"
+	"github.com/pingcap/errors"
+)
+
+// KeyStore tracks roles, keys, etc. and verifies signatures against this metadata.
+type KeyStore map[string]roleKeys
+
+type roleKeys struct {
+	threshold uint
+	// key id -> public key
+	keys map[string]crypto.PubKey
+}
+
+// NewKeyStore return a KeyStore
+func NewKeyStore() *KeyStore {
+	return &KeyStore{}
+}
+
+// AddKeys clears all keys for role, then adds all supplied keys and stores the threshold value.
+func (s *KeyStore) AddKeys(role string, threshold uint, keys map[string]*KeyInfo) error {
+	if threshold == 0 {
+		return errors.Errorf("invalid threshold (0)")
+	}
+
+	(*s)[role] = roleKeys{threshold: threshold, keys: map[string]crypto.PubKey{}}
+
+	for id, info := range keys {
+		pub, err := info.publicKey()
+		if err != nil {
+			return err
+		}
+		(*s)[role].keys[id] = pub
+	}
+
+	return nil
+}
+
+func newSignatureError(fname string, err error) *SignatureError {
+	return &SignatureError{
+		fname: fname,
+		err:   err,
+	}
+}
+
+// IsSignatureError check if the err is SignatureError.
+func IsSignatureError(err error) bool {
+	_, ok := err.(*SignatureError)
+	return ok
+}
+
+// SignatureError the signature of a file is incorrect.
+type SignatureError struct {
+	fname string
+	err   error
+}
+
+func (s *SignatureError) Error() string {
+	return fmt.Sprintf("invalid signature for file %s: %s", s.fname, s.err.Error())
+}
+
+// transitionRoot checks that signed is verified by signatures using newThreshold, and if so, updates the keys for the root
+// role in the key store.
+func (s *KeyStore) transitionRoot(signed []byte, newThreshold uint, signatures []Signature, newKeys map[string]*KeyInfo) error {
+	oldKeys := (*s)[ManifestTypeRoot]
+
+	err := s.AddKeys(ManifestTypeRoot, newThreshold, newKeys)
+	if err != nil {
+		return err
+	}
+
+	err = s.verifySignature(signed, ManifestTypeRoot, signatures, ManifestFilenameRoot)
+	if err != nil {
+		// Restore the old root keys.
+		(*s)[ManifestTypeRoot] = oldKeys
+		return err
+	}
+
+	return nil
+}
+
+// verifySignature verifies all supplied signatures are correct for signed. Also, that there are at least threshold signatures,
+// and that they all belong to the correct role.
+func (s *KeyStore) verifySignature(signed []byte, role string, signatures []Signature, filename string) error {
+	if s == nil {
+		return nil
+	}
+
+	// Check for duplicate signatures.
+	has := make(map[string]struct{})
+	for _, sig := range signatures {
+		if _, ok := has[sig.KeyID]; ok {
+			return errors.Errorf("signature section of %s contains duplicate signatures", filename)
+		}
+		has[sig.KeyID] = struct{}{}
+	}
+
+	keys, ok := (*s)[role]
+	if !ok {
+		return errors.Errorf("Unknown role %s", role)
+	}
+
+	var validSigs uint = 0
+	for _, sig := range signatures {
+		key, ok := keys.keys[sig.KeyID]
+		if !ok {
+			continue
```

Comment:
then this is not "all supplied signatures are correct for signed" ?

Reply:
I’ll change the comment. This is for handling root manifest old and new keys